### ISSUE TITLE
aws_s3_bucket_policy marked as modified on every apply regardless of changes

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-aws#596]
_This issue was originally opened by @kerin as hashicorp/terraform#12524. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

### Terraform Version
Terraform v0.8.8

### Affected Resource(s)
- aws_s3_bucket_policy

### Terraform Configuration Files
```hcl
resource "aws_s3_bucket_policy" "source" {
    bucket = "${aws_s3_bucket.source.id}"
    policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "DenyIncorrectEncryptionHeaderInSource",
      "Effect": "Deny",
      "Action": "s3:PutObject",
      "Resource": "${aws_s3_bucket.source.arn}/*",
      "Principal": "*",
      "Condition": {
        "StringNotEquals": {
          "s3:x-amz-server-side-encryption": "AES256"
        }
      }
    },
    {
      "Sid": "DenyUnEncryptedObjectUploadsInSource",
      "Effect": "Deny",
      "Action": "s3:PutObject",
      "Resource": "${aws_s3_bucket.source.arn}/*",
      "Principal": "*",
      "Condition": {
        "Null": {
          "s3:x-amz-server-side-encryption": true
        }
      }
    }
  ]
}
EOF
}
```

### Debug Output
https://gist.github.com/kerin/c27c23765dc0366147961b49059788c3

### Expected Behavior
Terraform should not see modifications to unchanged resources.

### Actual Behavior
Terraform thinks there's a modification to be applied on each `plan` or `apply`.

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform plan`
2. Observe planned changes to bucket policy
3. `terraform apply`
4. `terraform plan`
5. Observe that Terraform plans to apply the same modification again.

### Important Factoids
None of these attempts to work around the issue have had any effect:
- Removing all linebreaks and spaces from the policy document
- Using a separate JSON document via `$file()`
- Replacing references to resource ARNs with static values

**Comment by austinmills** [comment]
I am also running into this problem. Even using the aws_iam_policy_document suggested in https://github.com/hashicorp/terraform/issues/12524 still results in the policy being identified as a change during each run.

**Comment by bflad** [comment]
@austinmills what happens if you replace:

```
identifiers = ["127311923021"]
```

with:

```
identifiers = ["arn:aws:iam::127311923021:root"]
```

For reference, I have created an issue in the policy equivalence library upstream surrounding this behavior: https://github.com/jen20/awspolicyequivalence/issues/3

**Comment by austinmills** [comment]
@bflad That works! With the longer identifier format, it reports no difference. Thank you!

**Comment by bflad** [comment]
Hi folks! 👋 This issue was marked to close in #3832 due to a dependency update to an upstream library we use to compare the equivalence of two IAM policies: https://github.com/jen20/awspolicyequivalence/. The updated library should now treat account ID `123456789012` principals as equivalent to IAM account root ARNs `arn:PARTITION:iam::123456789012:root`. You can see some examples of what will now match in the upstream library unit testing:

https://github.com/jen20/awspolicyequivalence/blob/9fbcaca9f9f868b9560463d0790aae33b2322945/aws_policy_equivalence_test.go#L1076-L1200

This update will be released in v1.12.0 of the AWS provider, which we expect to release later this week.

I apologize this may not cover all IAM policy equivalence issues, but please do consider opening new Github issues in the https://github.com/terraform-providers/terraform-provider-aws/ or https://github.com/jen20/awspolicyequivalence/ for other specific examples that show as inequivalent.

**Comment by bflad** [comment]
This has been released in [version 1.12.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1120-march-23-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
